File: lib/services/cv_content_validator.dart
// Detecta uploads que claramente NÃO são currículos (extratos bancários,
// docs gov.br, holerites). Espelho da heurística no servidor
// (`supabase/functions/_shared/cv_content_validator.ts`) — manter os dois
// em paralelo: cliente avisa antes do upload pra UX boa, servidor rejeita
// como segunda camada.
//
// Conservador por design: nunca dispara em CV legítimo que apenas mencione
// um banco como experiência ou curso.

enum NonCvCategory { bankStatement, govIdDocument, payroll }

class NonCvDetection {
  final bool isNonCv;
  final NonCvCategory? category;
  final List<String> reasons;

  const NonCvDetection({
    required this.isNonCv,
    this.category,
    this.reasons = const [],
  });

  static const ok = NonCvDetection(isNonCv: false);
}

class CvContentValidator {
  static NonCvDetection detect(String rawText) {
    if (rawText.length < 50) return NonCvDetection.ok;
    final t = rawText.toLowerCase();

    // Extrato bancário: precisa de >=2 sinais específicos de movimentação.
    final bankSignals = <String>[];
    if (t.contains('saldo final')) bankSignals.add('saldo final');
    if (t.contains('saldo inicial')) bankSignals.add('saldo inicial');
    if (t.contains('total de entradas') ||
        t.contains('total de saídas') ||
        t.contains('total de saidas')) {
      bankSignals.add('total entradas/saídas');
    }
    if (t.contains('transferência recebida') ||
        t.contains('transferencia recebida')) {
      bankSignals.add('transferência recebida');
    }
    if (t.contains('transferência enviada') ||
        t.contains('transferencia enviada')) {
      bankSignals.add('transferência enviada');
    }
    if (t.contains('movimentações') || t.contains('movimentacoes')) {
      bankSignals.add('movimentações');
    }
    if (t.contains('pix') && t.contains('agência') && t.contains('conta')) {
      bankSignals.add('pix+agência+conta');
    }
    if (bankSignals.length >= 2) {
      return NonCvDetection(
        isNonCv: true,
        category: NonCvCategory.bankStatement,
        reasons: bankSignals,
      );
    }

    // Documento gov.br / identidade.
    final govSignals = <String>[];
    if (t.contains('gov.br')) govSignals.add('gov.br');
    if (t.contains('dados de pessoa física') ||
        t.contains('dados de pessoa fisica')) {
      govSignals.add('dados de pessoa física');
    }
    if (t.contains('nome da mãe') || t.contains('nome da mae')) {
      govSignals.add('nome da mãe');
    }
    if (t.contains('título de eleitor') || t.contains('titulo de eleitor')) {
      govSignals.add('título de eleitor');
    }
    if (t.contains('situação cadastral') ||
        t.contains('situacao cadastral')) {
      govSignals.add('situação cadastral');
    }
    if (t.contains('assinado digitalmente por gov.br') ||
        t.contains('validar.iti.gov.br')) {
      govSignals.add('assinatura digital gov.br');
    }
    if (govSignals.length >= 2) {
      return NonCvDetection(
        isNonCv: true,
        category: NonCvCategory.govIdDocument,
        reasons: govSignals,
      );
    }

    // Holerite / contracheque. Exige ≥2 sinais pra não bloquear CV de
    // RH/DP que mencione "holerite" ou "contracheque" como experiência.
    final payrollSignals = <String>[];
    if (t.contains('holerite')) payrollSignals.add('holerite');
    if (t.contains('contracheque') || t.contains('contra-cheque')) {
      payrollSignals.add('contracheque');
    }
    if (t.contains('recibo de pagamento de salário') ||
        t.contains('recibo de pagamento de salario')) {
      payrollSignals.add('recibo de pagamento de salário');
    }
    if (t.contains('fgts') &&
        t.contains('inss') &&
        (t.contains('proventos') || t.contains('descontos'))) {
      payrollSignals.add('fgts+inss+proventos/descontos');
    }
    if ((t.contains('valor líquido') || t.contains('valor liquido')) &&
        t.contains('competência')) {
      payrollSignals.add('valor líquido+competência');
    }
    if (payrollSignals.length >= 2) {
      return NonCvDetection(
        isNonCv: true,
        category: NonCvCategory.payroll,
        reasons: payrollSignals,
      );
    }

    return NonCvDetection.ok;
  }

  static String messageFor(NonCvCategory category) {
    switch (category) {
      case NonCvCategory.bankStatement:
        return 'Isso parece um extrato bancário, não um currículo. Envie o seu CV.';
      case NonCvCategory.govIdDocument:
        return 'Isso parece um documento de identidade, não um currículo. Envie o seu CV.';
      case NonCvCategory.payroll:
        return 'Isso parece um holerite, não um currículo. Envie o seu CV.';
    }
  }
}

File: supabase/functions/_shared/cv_content_validator.ts
// Detecta uploads que claramente NÃO são currículos (extratos bancários,
// documentos de identidade gov.br, holerites). Conservador: prefere deixar
// passar do que rejeitar CV legítimo. Cada categoria exige uma combinação
// de sinais específicos — palavra solta não dispara.
//
// Casos reais observados em produção que motivaram esse validador:
//   - Extrato Nubank de 14k chars com dados financeiros + dados de terceiros
//   - Documento gov.br com CPF, nome da mãe, endereço
//
// Falsos positivos a evitar: CVs que mencionam bancos como experiência
// profissional ou cursos ("Excel Básico - Santander", "Trabalhou no Bradesco").
// Por isso a heurística NUNCA dispara só por nome de banco.

export type NonCvCategory = 'bank_statement' | 'gov_id_document' | 'payroll'

export interface NonCvDetection {
  isNonCv: boolean
  category: NonCvCategory | null
  reasons: string[]
}

export function detectNonCvContent(rawText: string): NonCvDetection {
  if (!rawText || rawText.length < 50) {
    return { isNonCv: false, category: null, reasons: [] }
  }

  const t = rawText.toLowerCase()

  // ── Extrato bancário ──────────────────────────────────────────────────
  // Exige pelo menos 2 sinais específicos de movimentação financeira.
  const bankSignals: string[] = []
  if (t.includes('saldo final')) bankSignals.push('saldo final')
  if (t.includes('saldo inicial')) bankSignals.push('saldo inicial')
  if (t.includes('total de entradas') || t.includes('total de saídas') || t.includes('total de saidas')) {
    bankSignals.push('total entradas/saídas')
  }
  if (t.includes('transferência recebida') || t.includes('transferencia recebida')) {
    bankSignals.push('transferência recebida')
  }
  if (t.includes('transferência enviada') || t.includes('transferencia enviada')) {
    bankSignals.push('transferência enviada')
  }
  if (t.includes('movimentações') || t.includes('movimentacoes')) bankSignals.push('movimentações')
  // "pix" + "agência" + "conta" juntos = quase certeza de extrato
  if (t.includes('pix') && t.includes('agência') && t.includes('conta')) {
    bankSignals.push('pix+agência+conta')
  }
  if (bankSignals.length >= 2) {
    return {
      isNonCv: true,
      category: 'bank_statement',
      reasons: bankSignals,
    }
  }

  // ── Documento gov.br / identidade ─────────────────────────────────────
  // "gov.br" sozinho é fraco (pode ser link). Exige combinação com campo
  // típico de doc identidade.
  const govSignals: string[] = []
  if (t.includes('gov.br')) govSignals.push('gov.br')
  if (t.includes('dados de pessoa física') || t.includes('dados de pessoa fisica')) {
    govSignals.push('dados de pessoa física')
  }
  if (t.includes('nome da mãe') || t.includes('nome da mae')) govSignals.push('nome da mãe')
  if (t.includes('título de eleitor') || t.includes('titulo de eleitor')) {
    govSignals.push('título de eleitor')
  }
  if (t.includes('situação cadastral') || t.includes('situacao cadastral')) {
    govSignals.push('situação cadastral')
  }
  if (t.includes('assinado digitalmente por gov.br') || t.includes('validar.iti.gov.br')) {
    govSignals.push('assinatura digital gov.br')
  }
  if (govSignals.length >= 2) {
    return {
      isNonCv: true,
      category: 'gov_id_document',
      reasons: govSignals,
    }
  }

  // ── Holerite / contracheque ───────────────────────────────────────────
  // Exige ≥2 sinais (igual extrato bancário e gov.br) pra não bloquear
  // CV de RH/DP que mencione "holerite" ou "contracheque" como parte da
  // descrição de experiência profissional.
  const payrollSignals: string[] = []
  if (t.includes('holerite')) payrollSignals.push('holerite')
  if (t.includes('contracheque') || t.includes('contra-cheque')) payrollSignals.push('contracheque')
  if (t.includes('recibo de pagamento de salário') || t.includes('recibo de pagamento de salario')) {
    payrollSignals.push('recibo de pagamento de salário')
  }
  // FGTS + INSS + proventos = padrão de holerite
  if (t.includes('fgts') && t.includes('inss') && (t.includes('proventos') || t.includes('descontos'))) {
    payrollSignals.push('fgts+inss+proventos/descontos')
  }
  // Valor líquido + competência = padrão de holerite
  if ((t.includes('valor líquido') || t.includes('valor liquido')) && t.includes('competência')) {
    payrollSignals.push('valor líquido+competência')
  }
  if (payrollSignals.length >= 2) {
    return {
      isNonCv: true,
      category: 'payroll',
      reasons: payrollSignals,
    }
  }

  return { isNonCv: false, category: null, reasons: [] }
}

export function nonCvMessage(category: NonCvCategory): string {
  switch (category) {
    case 'bank_statement':
      return 'Isso parece um extrato bancário, não um currículo. Envie o seu CV.'
    case 'gov_id_document':
      return 'Isso parece um documento de identidade, não um currículo. Envie o seu CV.'
    case 'payroll':
      return 'Isso parece um holerite, não um currículo. Envie o seu CV.'
  }
}
